Implement create todo endpoint

yamin-faruque-sazim opened the pull request:
**DESCRIPTION OF CHANGES**

- Refactored MikroORM setup slightly to follow best practices
- Created Todo entity
- Created the necessary files for the endpoint : Controller, Service, Repository 
- Added basic validation via class validators 

https://github.com/user-attachments/assets/efe4ee28-bdd2-45f6-944a-8df3af750f11

yamin-faruque-sazim committed:
REFACTOR(FTDA): Change MikroORM config
CHORE(FTDA):Moved MikroORM connection setup from main.ts to app.module.ts
FEAT(FTDA):Define Todo entity
FEAT(FTDA):Implement an endpoint to create a todo object and have it persist in the DB

yamin-faruque-sazim committed:
FEAT(FTDA): Add some basic validation via class-validators

ishmam-iqbal-sazim commented on src/common/entities/todo.entity.ts:
```diff
@@ -0,0 +1,25 @@
+  description!: string;
+
+  @Property({ default: false })
+  isCompleted: boolean = false;
```
```suggestion
  @Property({ default: false, fieldName: "is_completed" })
  isCompleted: boolean = false;
```

We should add fieldName (in snake_case) explicitly for all of the attributes.

ishmam-iqbal-sazim commented on src/db/migrations/Migration20240904172648.ts:
In this particular case, multiple migrations are not necessary. We can just consolidate the migrations into one since you are just working with one table.

ishmam-iqbal-sazim commented on src/db/migrations/Migration20240905070313.ts:
Same comment as above

ishmam-iqbal-sazim commented on tsconfig.json:
```diff
@@ -16,6 +16,7 @@
     "forceConsistentCasingInFileNames": false,
-    "noFallthroughCasesInSwitch": false
+    "noFallthroughCasesInSwitch": false,
+    "esModuleInterop": true
```
Please add `strict: true` to the tsconfig. 
It'd actually be best if you can just copy the tsconfig from the backend template and paste it here.

ishmam-iqbal-sazim commented on src/todos/create-todo.dto.ts:
```diff
@@ -0,0 +1,22 @@
+export class CreateTodoDto {
+  @IsNotEmpty()
+  @IsString()
+  title: string;
```
```suggestion
  title!: string;
```

Best to add assertion operator as we need to let typescript know that this will always be assigned a value. Same comment for all dto attributes that are mandatory

ishmam-iqbal-sazim commented on src/common/entities/todo.entity.ts:
```diff
@@ -0,0 +1,25 @@
+import { ETodoPriority } from '../enums/todo-priority.enum';
+import { TodoRepository } from '../../todos/todos.repository';
+
+@Entity({ repository: () => TodoRepository })
```
Should add tableName as well

ishmam-iqbal-sazim commented on src/common/entities/todo.entity.ts:
```diff
@@ -0,0 +1,25 @@
+
+@Entity({ repository: () => TodoRepository })
+export class Todo {
+  @PrimaryKey({ type: 'uuid' })
```
This is missing the type`[EntityRepositoryType]?: repo`

ishmam-iqbal-sazim commented on src/common/entities/todo.entity.ts:
```diff
@@ -0,0 +1,25 @@
+import { TodoRepository } from '../../todos/todos.repository';
+
+@Entity({ repository: () => TodoRepository })
+export class Todo {
+  @PrimaryKey({ type: 'uuid' })
+  id: string = uuidv4();
+
+  @Property()
+  title!: string;
+
+  @Property({ type: 'text' })
+  description!: string;
+
+  @Property({ default: false })
+  isCompleted: boolean = false;
+
+  @Enum({ items: () => ETodoPriority })
+  priority!: ETodoPriority;
+
+  @Property({ type: 'date' })
+  dueDate!: Date;
+}
```
This class is missing the type`[EntityRepositoryType]?: repo`

ishmam-iqbal-sazim commented on src/todos/todos.module.ts:
```diff
@@ -0,0 +1,13 @@
+@Module({
+  imports: [MikroOrmModule.forFeature([Todo])],
+  controllers: [TodosController],
+  providers: [TodosService, TodoRepository],
```
```suggestion
  providers: [TodosService],
```

You don't need to provide repository if you add the repository to the entity. Importing just the entity should be fine

ishmam-iqbal-sazim commented on src/todos/todos.service.ts:
```diff
@@ -0,0 +1,14 @@
+export class TodosService {
+  constructor(private todoRepository: TodoRepository) {}
+
+  async createTodos(createTodoDto: CreateTodoDto): Promise<Todo> {
```
```suggestion
  async createOne(createOneDto: CreateOneDto): Promise<Todo> {
```

This is in the todosService so naming it createOne should be fine as you would be using this method like so: `this.todosService.createOne`

ishmam-iqbal-sazim commented on src/todos/todos.repository.ts:
```diff
@@ -0,0 +1,17 @@
+  constructor(em: EntityManager) {
+    super(em, Todo);
+  }
+  async createTodo(createTodoDto: CreateTodoDto): Promise<Todo> {
```
```suggestion
  async createOne(createOneDto: CreateOneDto): Promise<Todo> {
```

ishmam-iqbal-sazim commented on src/todos/todos.repository.ts:
```diff
@@ -0,0 +1,17 @@
+    const todo = this.create(createTodoDto);
+    await this.em.persistAndFlush(todo);
+    return todo;
+  }
```
You probably don't need this method.
You could just call the create and persistAndFlush methods in the service layer.

yamin-faruque-sazim committed:
CHORE(FTDA): Address PR comments
- Refactored `Todo` entity: added `tableName`, set `fieldName` for attributes, and included `[EntityRepositoryType]`.
- Consolidated migrations into one as suggested.
- Updated DTO properties with assertion operators for mandatory fields.
- Modified `TodosService` to rename method to `createOne` for consistency.
- Removed redundant repository injection in `TodosModule`.
- Simplified `createTodo` logic by directly handling create and persist in the service layer.
- Enhanced `tsconfig.json` with strict settings as per review suggestions.

ishmam-iqbal-sazim commented on tsconfig.json:
```diff
@@ -6,16 +29,60 @@
+//       "@/test/*": ["../test/*"]
+//     }
+//   }
+// }
```
Let's add a new line here

ishmam-iqbal-sazim commented on tsconfig.json:
```diff
@@ -1,3 +1,26 @@
+// {
```
Let's remove all the comments in this file

ishmam-iqbal-sazim reviewed (approved):
Approved pending comments
